File: leetcode332/leetcode332_1.cpp
class Solution {
public:
    vector<string> findItinerary(vector<vector<string>>& tickets) {
        for (auto ticket : tickets) {
            targets[ticket[0]].insert(ticket[1]);
        }
        
        visit("JFK");
        return vector<string> (route.rbegin(), route.rend());
    }
    
private:
    void visit(string airport) {
        while (targets[airport].size()) {
            string next = *targets[airport].begin();
            targets[airport].erase(targets[airport].begin());
            visit(next);
        }
        
        route.push_back(airport);
    }
    
    map<string, multiset<string>> targets;
    vector<string> route;
};


// Runtime: 24 ms, faster than 83.20% of C++ online submissions for Reconstruct Itinerary.
// Memory Usage: 17.5 MB, less than 25.00% of C++ online submissions for Reconstruct Itinerary.
// time complexity: O(n)
// space complexity: O(n)
// https://leetcode.com/problems/reconstruct-itinerary/discuss/78768/Short-Ruby-Python-Java-C%2B%2B

// class Solution {
// public:
//     vector<string> findItinerary(vector<vector<string>>& tickets) {
//         for (auto ticket : tickets)
//             targets[ticket[0]].insert(ticket[1]);
//         visit("JFK");
//         return vector<string>(route.rbegin(), route.rend());
//     }

//     map<string, multiset<string>> targets;
//     vector<string> route;

//     void visit(string airport) {
//         while (targets[airport].size()) {
//             string next = *targets[airport].begin();
//             targets[airport].erase(targets[airport].begin());
//             visit(next);
//         }
//         route.push_back(airport);
//     }
// };

File: leetcode332/leetcode332_1.java
class Solution {
    public List<String> findItinerary(List<List<String>> tickets) {
        Map<String, PriorityQueue<String>> targets = new HashMap<>();
        List<String> route = new LinkedList();

        for (List<String> ticket : tickets)
            targets.computeIfAbsent(ticket.get(0), k -> new PriorityQueue()).add(ticket.get(1));
        visit(targets, route, "JFK");
        return route;
    }
    
    

    void visit(Map<String, PriorityQueue<String>> targets, List<String> route, String airport) {
        while(targets.containsKey(airport) && !targets.get(airport).isEmpty())
            visit(targets, route, targets.get(airport).poll());
        route.add(0, airport);
    }
}

// Runtime: 5 ms, faster than 76.49% of Java online submissions for Reconstruct Itinerary.
// Memory Usage: 39.2 MB, less than 94.18% of Java online submissions for Reconstruct Itinerary.
// time Complexity: O(n)
// space complexity: O(n)
// https://leetcode.com/problems/reconstruct-itinerary/discuss/78768/Short-Ruby-Python-Java-C%2B%2B